src/compute-types/src/plan/lowering.rs

```diff
@@ -380,37 +383,134 @@ impl Context {
                     b_keys,
                 )
             }
-            MirRelationExpr::FlatMap { input, func, exprs } => {
-                let (input, keys) = self.lower_mir_expr(input)?;
-                // This stage can absorb arbitrary MFP instances.
-                let mfp = mfp.take();
-                let mut exprs = exprs.clone();
-                let input_key = if let Some((k, permutation, _)) = keys.arbitrary_arrangement() {
-                    // We don't permute the MFP here, because it runs _after_ the table function,
-                    // whose output is in a fixed order.
-                    //
-                    // We _do_, however, need to permute the `expr`s that provide input to the
-                    // `func`.
-                    for expr in &mut exprs {
-                        expr.permute_map(permutation);
+            MirRelationExpr::FlatMap {
```

Comment:
At a high level, this is more of a physical optimization, which are sometimes unpleasant to make a part of the MIR pipeline.

The specific problem here with putting this into the MIR pipeline would be that we'd need to modify MIR's semantics: MIR's `Reduce` currently always emits exactly 1 row per group, but the fused `Reduce-FlatMap` can emit multiple rows per group. Such semantic changes of MIR are very scary, since various parts of the optimizer assume that `Reduce` emits only 1 row per group, and it would be very hard to hunt down all these parts. (For example, key inference infers the group key as a unique key.)

(Btw. the MIR pipeline currently has a physical part, where the most important thing is `JoinImplementation`, but we are planning to move also `JoinImplementation` to the MIR-to-LIR lowering.)

Reply:
Thanks, that makes sense.